services/hierarchy/src/main/kotlin/OrphanRemovalService.kt

```diff
@@ -86,220 +94,209 @@ class OrphanRemovalService(
-            RemoteArtifactsTable.deleteWhere {
-                id notInSubQuery (
+        RemoteArtifactsTable.deleteWhereNotExists { id ->
+            val subQuery = PackagesTable
+                .select(PackagesTable.binaryArtifactId.alias("id"))
+                .union(
                     PackagesTable
-                        .select(PackagesTable.binaryArtifactId.alias("id"))
-                        .where(PackagesTable.binaryArtifactId.isNotNull())
-                        .union(
-                            PackagesTable
-                                .select(PackagesTable.sourceArtifactId.alias("id"))
-                                .where(PackagesTable.sourceArtifactId.isNotNull())
-                        )
-                        .union(
-                            PackageProvenancesTable
-                                .select(PackageProvenancesTable.artifactId.alias("id"))
-                                .where(PackageProvenancesTable.artifactId.isNotNull())
-                        )
-                        .union(
-                            PackageCurationDataTable
-                                .select(PackageCurationDataTable.binaryArtifactId.alias("id"))
-                                .where(PackageCurationDataTable.binaryArtifactId.isNotNull())
-                        )
-                        .union(
-                            PackageCurationDataTable
-                                .select(PackageCurationDataTable.sourceArtifactId.alias("id"))
-                                .where(PackageCurationDataTable.sourceArtifactId.isNotNull())
-                        )
-                        .union(
-                            SnippetsTable
-                                .select(SnippetsTable.artifactId.alias("id"))
-                                .where(SnippetsTable.artifactId.isNotNull())
-                        )
-                    )
-            }
+                        .select(PackagesTable.sourceArtifactId)
+                )
+                .union(
+                    PackageProvenancesTable
+                        .select(PackageProvenancesTable.artifactId)
+                )
+                .union(
+                    PackageCurationDataTable
+                        .select(PackageCurationDataTable.binaryArtifactId)
+                )
+                .union(
+                    PackageCurationDataTable
+                        .select(PackageCurationDataTable.sourceArtifactId)
+                )
+                .union(
+                    SnippetsTable
+                        .select(SnippetsTable.artifactId)
+                )
+
+            unionCondition(subQuery, id)
         }
 }
+
+/**
+ * Add a condition to match the given [matchColumn] to the given [unionQuery]. This results in a statement of the
+ * form: SELECT * FROM (SELECT * FROM sub_query) WHERE sub_query.id = matchColumn.
+ * TODO: Check whether there is a better way to do this with Exposed.
```

Comment:
Maybe it's possible to pass the return value of the `alias` call, e.g., `PackagesTable.binaryArtifactId.alias("id")`, to this function and use it directly in the `where` condition.

Reply:
I have tested this, but it does not work. The alias obviously contains the table name; so the condition that is generated would have the form (in your example): `packages.id = vcs_info.id`.